In [ ]:
Análisis de Datos - Luis Carlos Barraza Terraza - 19.768.492 Grupo 202016908_50 

# Análisis con KNN (Vecinos más cercanos):

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [9]:
df = pd.read_csv("/Users/luisk/Desktop/Ej2-Cleaned-Data.csv")
data = df
data

Fever  Tiredness  Dry-Cough  Difficulty-in-Breathing  Sore-Throat  \
0           1          1          1                        1            1   
1           1          1          1                        1            1   
2           1          1          1                        1            1   
3           1          1          1                        1            1   
4           1          1          1                        1            1   
...       ...        ...        ...                      ...          ...   
316795      0          0          0                        0            0   
316796      0          0          0                        0            0   
316797      0          0          0                        0            0   
316798      0          0          0                        0            0   
316799      0          0          0                        0            0   

        None_Sympton  Pains  Nasal-Congestion  Runny-Nose  Diarrhea  ...  \
0                  0      1                 1           1         1  ...   
1                  0      1                 1           1         1  ...   
2                  0      1                 1           1         1  ...   
3                  0      1                 1           1         1  ...   
4                  0      1                 1           1         1  ...   
...              ...    ...               ...         ...       ...  ...   
316795             1      0                 0           0         0  ...   
316796             1      0                 0           0         0  ...   
316797             1      0                 0           0         0  ...   
316798             1      0                 0           0         0  ...   
316799             1      0                 0           0         0  ...   

        Gender_Male  Gender_Transgender  Severity_Mild  Severity_Moderate  \
0                 1                   0              1                  0   
1                 1                   0              1                  0   
2                 1                   0              1                  0   
3                 1                   0              0                  1   
4                 1                   0              0                  1   
...             ...                 ...            ...                ...   
316795            0                   1              0                  0   
316796            0                   1              0                  0   
316797            0                   1              0                  0   
316798            0                   1              0                  0   
316799            0                   1              0                  0   

        Severity_None  Severity_Severe  Contact_Dont-Know  Contact_No  \
0                   0                0                  0           0   
1                   0                0                  0           1   
2                   0                0                  1           0   
3                   0                0                  0           0   
4                   0                0                  0           1   
...               ...              ...                ...         ...   
316795              0                1                  0           1   
316796              0                1                  1           0   
316797              1                0                  0           0   
316798              1                0                  0           1   
316799              1                0                  1           0   

        Contact_Yes  Country  
0                 1    China  
1                 0    China  
2                 0    China  
3                 1    China  
4                 0    China  
...             ...      ...  
316795            0    Other  
316796            0    Other  
316797            1    Other  
316798            0    Other  
316799            0    Othe

# Eliminamos la columna 'Country' ya que no la necesitas para el modelo

In [10]:
data = data.drop(['Country'], axis=1)

# Divide el conjunto de datos en características (X) y etiquetas (y)

In [12]:
X = data[['Tiredness', 'Dry-Cough', 'Difficulty-in-Breathing', 'Sore-Throat',  'Pains', 'Runny-Nose', 'Nasal-Congestion', 'Fever']]
y = data['Severity_Moderate']


# Divide el conjunto de datos en conjuntos de entrenamiento y prueba

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Inicializamos el clasificador KNN

In [16]:
knn_classifier = KNeighborsClassifier(n_neighbors=5)

# Entrenamos el modelo

In [17]:
knn_classifier.fit(X_train, y_train)

KNeighborsClassifier()

# Realizamos predicciones en el conjunto de prueba

In [18]:
y_pred_knn = knn_classifier.predict(X_test)

# Evalúamos el rendimiento del modelo

In [21]:
accuracy_knn = accuracy_score(y_test, y_pred_knn)
print(f'Accuracy (KNN): {accuracy_knn}')

Accuracy (KNN): 0.7112058080808081


# Otros resultados

In [ ]:
print('\nClassification Report:')
print(classification_report(y_test, y_pred_knn))
print('\nConfusion Matrix:')
print(confusion_matrix(y_test, y_pred_knn))

# Análisis con Naive Bayes:

# Importamos las librerías necesarias

In [22]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import LabelEncoder

# Dividimos el conjunto de datos en características (X) y etiquetas (y)

In [24]:
label_encoder = LabelEncoder()
data['Severity_Moderate'] = label_encoder.fit_transform(data['Severity_Moderate'])

# Dividimos el conjunto de datos en conjuntos de entrenamiento y prueba

In [25]:
X = data.drop(['Severity_Moderate'], axis=1)
y = data['Severity_Moderate']

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Inicializamos el clasificador Naive Bayes

In [27]:
nb_classifier = MultinomialNB()

# Entrenamos el modelo

In [28]:
nb_classifier.fit(X_train, y_train)

MultinomialNB()

# Realizamos predicciones en el conjunto de prueba

In [29]:
y_pred_nb = nb_classifier.predict(X_test)

# Evaluamos el rendimiento del modelo

In [30]:
accuracy_nb = accuracy_score(y_test, y_pred_nb)
print(f'\nAccuracy (Naive Bayes): {accuracy_nb}')


Accuracy (Naive Bayes): 0.7998579545454545


# Otros resultados

In [31]:
print('\nClassification Report:')
print(classification_report(y_test, y_pred_nb))
print('\nConfusion Matrix:')
print(confusion_matrix(y_test, y_pred_nb))


Classification Report:
              precision    recall  f1-score   support

           0       0.79      1.00      0.88     47451
           1       1.00      0.20      0.34     15909

    accuracy                           0.80     63360
   macro avg       0.89      0.60      0.61     63360
weighted avg       0.84      0.80      0.75     63360


Confusion Matrix:
[[47451     0]
 [12681  3228]]


#KNN (Vecinos más cercanos):

Precisión (Accuracy): El modelo acertó alrededor del 71.12% de las veces en predecir si un paciente tenía una infección moderada o no.

Precision: De las veces que el modelo dijo que era una infección moderada, acertó aproximadamente el 71%.

Recall: De todos los casos de infección moderada reales, el modelo identificó correctamente alrededor del 80%.

F1-score: Una combinación de precision y recall, alrededor del 75%.
Confusion Matrix:

Verdaderos Positivos (VP): 24192 pacientes clasificados correctamente como moderadamente infectados.

Verdaderos Negativos (VN): 38467 pacientes clasificados correctamente como no moderadamente infectados.

Falsos Positivos (FP): 5069 pacientes clasificados incorrectamente como moderadamente infectados.

Falsos Negativos (FN): 11632 pacientes clasificados incorrectamente como no moderadamente infectados.



Naive Bayes:

Precisión (Accuracy): El modelo acertó alrededor del 79.99% de las veces en predecir si un paciente tenía una infección moderada o no.



Precision: De las veces que el modelo dijo que era una infección moderada, acertó aproximadamente el 79%.
Recall: De todos los casos de infección moderada reales, el modelo identificó correctamente el 100%.
F1-score: Una combinación de precision y recall, alrededor del 88%.
Confusion Matrix:

Verdaderos Positivos (VP): 3228 pacientes clasificados correctamente como moderadamente infectados.

Verdaderos Negativos (VN): 47451 pacientes clasificados correctamente como no moderadamente infectados.

Falsos Positivos (FP): 0 pacientes clasificados incorrectamente como moderadamente infectados.

Falsos Negativos (FN): 12681 pacientes clasificados incorrectamente como no moderadamente infectados.



En términos sencillos, ambos modelos predicen si un paciente tiene una infección moderada o no. El modelo Naive Bayes tuvo un rendimiento generalmente mejor, especialmente en identificar correctamente los casos reales de infección moderada. La matriz de confusión nos dice cuántas veces el modelo acertó o se equivocó en sus predicciones. En resumen, el modelo Naive Bayes parece ser más confiable para este escenario específico.